Traverse dependencies and dev dependencies separately

jonahwilliams opened the pull request:
Fixes #15080

Currently, `flutter update-packages --force-upgrade` will combine the dependencies and dev_dependencies sections of a pubspec before retrieving the set of transitive dependencies.  All of the resulting deps are then added to the dev dependency section.

This behavior is not quite correct, as documented in #14992.  The transitive dependencies of the package direct dependencies must be included in the dependencies section.  For example, the main flutter package (currently) has a dep on `http`.  We should expect the pubspec to look something like this:

```
dependencies:
  http: 0.11.3+16
  ...
  async: 2.0.6 # TRANSITIVE DEPENDENCY
  charcode: 1.1.1 # TRANSITIVE DEPENDENCY
  http_parser: 3.1.1 # TRANSITIVE DEPENDENCY
  path: 1.5.1 # TRANSITIVE DEPENDENCY
  source_span: 1.4.0 # TRANSITIVE DEPENDENCY
  string_scanner: 1.0.2 # TRANSITIVE DEPENDENCY
```
Instead  It looks something like the following, with the transitive deps of `package:http` added in dev_dependencies:

```
dependencies:
  http: 0.11.3+16
  ...

dev_dependencies:
  async: 2.0.6 # TRANSITIVE DEPENDENCY
  charcode: 1.1.1 # TRANSITIVE DEPENDENCY
  http_parser: 3.1.1 # TRANSITIVE DEPENDENCY
  path: 1.5.1 # TRANSITIVE DEPENDENCY
  source_span: 1.4.0 # TRANSITIVE DEPENDENCY
  string_scanner: 1.0.2 # TRANSITIVE DEPENDENCY
```

To correct this process, I've updated the `flutter update-packages` command to separate the dependencies and dev_dependenices, then to do two traversals of the resulting transitive deps.  The same `seen` set is used between both traversals to prevent repeats, since any package in the dependencies section is implicitly a dev dependency too.  Furthermore, when determining the dependencies of a PubspecYaml, we need to distinguish between regular and dev deps too

I also ran the updated command on the flutter repos

Commit:
traverse dependencies and dev dependencies separately

Commit:
separate dev and reg deps

Commit:
update pubspecs

jonahwilliams committed:
Merge branch 'master' into pin_even_more

jonahwilliams commented:
The use of package:intl_translation seems to be adding a lot of unfortunate dependencies

Hixie commented:
While you're here, can you do me a favour and change the text "TRANSITIVE DEPENDENCY" to something like `THIS LINE IS AUTOGENERATED - TO UPDATE USE "flutter update-packages --force-upgrade` ? It might reduce the number of times that people try to manually update these lines.

Hixie commented:
For super duper bonus points (not at all required for this PR unless you feel motivated to do it), what you could do to REALLY stop people from hand-editing these is include a value at the bottom of each list which is some sort of hash of the values above, along with a comment saying to run the script to update it, and then add a test that opens each pubspec.yaml and verifies that the hashes match the currently listed dependencies.

Hixie commented on packages/flutter_tools/lib/src/commands/update_packages.dart:
```diff
@@ -398,6 +398,8 @@ class PubspecYaml {
           // We're in a section we care about. Try to parse out the dependency:
           final PubspecDependency dependency = PubspecDependency.parse(line, filename: filename);
           if (dependency != null) { // We got one!
+            // track whether or not this a dev dependency.
```
trivial nit: comments should start with a capital letter according to the dart style guide (which we include transitively through our style guide)

Hixie commented on packages/flutter_tools/lib/src/commands/update_packages.dart:
```diff
@@ -534,11 +550,18 @@ class PubspecYaml {
             }
             // Since we're in one of the places where we can list dependencies,
             // remember this as the current last known valid place to insert our
             // transitive dependencies.
```
this comment needs updating since you now insert things in two places

Hixie commented on packages/flutter_tools/lib/src/commands/update_packages.dart:
```diff
@@ -558,32 +581,52 @@ class PubspecYaml {
     // Only if there were no dependencies: or dev_dependencies: sections could
     // we get here with this still null, and we should not have any such files
```
this -> these

Hixie commented on packages/flutter_tools/lib/src/commands/update_packages.dart:
```diff
@@ -558,32 +581,52 @@ class PubspecYaml {
+    final Set<String> transitiveDevDependencies = new Set<String>();
+    
     // Merge the list of dependencies we've seen in this file and the dependencies we know this
     // file mentions that are already pinned (and which didn't get special processing above).
```
all the comments here (this one, the next one, the next one after that, etc) all need updating to make sense now that there's two lists

Hixie commented on packages/flutter_tools/lib/src/commands/update_packages.dart:
```diff
@@ -483,7 +493,10 @@ class PubspecYaml {
     Section section = Section.other; // the section we're currently handling
+
+    int endOfDirectDependencies;
     int lastPossiblePlace; // the line number where we're going to insert the transitive dependencies
```
probably worth renaming this one to be consistent with the one you added (also, comment needs updating)

Hixie commented:
![LGTM](http://junkyard.damowmow.com/577)

This is fantastic, thanks.

I didn't check the comments that aren't mentioned in the diff, but I recommend doing one last scan through the file to make sure all the comments still make sense.

tvolkert commented on examples/flutter_view/pubspec.yaml:
```diff
@@ -5,55 +5,17 @@ dependencies:
   flutter:
     sdk: flutter
 
-  args: 1.4.1 # TRANSITIVE DEPENDENCY
```
I wonder: how were these transitive deps here to begin with when there are no direct dev dependencies here?

Commit:
add checksum string and update transitive dep string.  Add additional verify command to check checksum

Commit:
handle merge conflict

Commit:
fix more conflicts

Commit:
cleanup comments

jonahwilliams commented:
@Hixie I added an initial implementation of a simple checksum over the package names + versions, and updated the magic string.  Also addressed comment comments.

Hixie commented on packages/flutter_tools/lib/src/commands/update_packages.dart:
```diff
@@ -312,7 +357,10 @@ enum DependencyKind {
 /// dependencies so that we can ignore them the next time we parse the
 /// pubspec.yaml file.
-const String kTransitiveMagicString = '# TRANSITIVE DEPENDENCY';
+const String kTransitiveMagicString= '# THIS LINE IS AUTOGENERATED - TO UPDATE USE "flutter update-packages --force-upgrade';
```
nit: missing a trailing `"` in the string, also, you have a space missing before the `=`.

jonahwilliams replied:
Fixed

Hixie commented:
![LGTM](http://junkyard.damowmow.com/577)

Hixie commented on packages/flutter_tools/lib/src/commands/update_packages.dart:
```diff
@@ -59,6 +59,11 @@ class UpdatePackagesCommand extends FlutterCommand {
         defaultsTo: false,
+      )
+      ..addFlag(
+        'verify-only',
```
call this from the test script? (`/dev/bots/test.dart`)

jonahwilliams replied:
Done

tvolkert reviewed (approved):
LGTM

This is awesome.

Commit:
Merge branch 'master' of github.com:flutter/flutter into pin_even_more

Commit:
add missing space, quote, run verify-onlu from bot script

Commit:
rerun tool with fixed string

jonahwilliams committed:
ensure correct working directory and bump versions

jonahwilliams committed:
flutter not flutterRoot
